mailpoet/lib/Automation/Engine/Builder/UpdateWorkflowController.php

```diff
@@ -52,6 +58,7 @@ public function updateWorkflow(int $id, array $data): Workflow {
     }
 
     if ($changed) {
+      $this->hooks->doWorkflowBeforeSave($workflow);
```

Comment:
Hmm, not sure. I'm kind of thinking I'd like to have the storage classes very simple and kind of low-level, and maybe that it should be the controller on a higher level to add events like these. But I'm not 100% convinced either way.

I see a difference here from things like Doctrine events (where you really want to run something on every DB update) because this hook (in my thinking) was more like "call it exactly once and only when the workflow is explicitly updated through the update API" and it seemed to me that `UpdateWorkflowController` could be the right service with that responsibility.

Ultimately, we may also need a lower-level "update" that will not trigger this (e.g., something is updated by a cron, not by a user), and therefore I kind of think the storage classes could try to stay low-level.

Reply:
That makes perfect sense to me, thank you for the explanation!